omnipaxos_core/src/sequence_paxos/leader.rs

```diff
@@ -48,19 +48,8 @@ where
             self.leader_state.set_promise(my_promise, self.pid, true);
             /* initialise longest chosen sequence and update state */
             self.state = (Role::Leader, Phase::Prepare);
-            let prep = Prepare {
-                n,
-                decided_idx,
-                n_accepted: self.internal_storage.get_accepted_round(),
-                accepted_idx,
-            };
-            /* send prepare */
-            for pid in &self.peers {
-                self.outgoing.push(PaxosMessage {
-                    from: self.pid,
-                    to: *pid,
-                    msg: PaxosMsg::Prepare(prep),
-                });
+            for pid in self.peers.clone() {
+                self.send_prepare(pid);
```

Comment:
Why cloning? I think it should work by dereferencing `*pid` like we did before.

Reply:
`send_prepare` requires a mutable reference to self since it mutates `self.outgoing`. The compiler is worried that we immutably borrow `self.peers` and then do a mutable borrow since it can't tell that we wont mutate `self.peers` in the `send_prepare` function call. https://smallcultfollowing.com/babysteps/blog/2018/11/01/after-nll-interprocedural-conflicts/
```
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable
  --> omnipaxos_core/src/sequence_paxos/leader.rs:52:17
   |
51 |             for pid in &self.peers {
   |                        -----------
   |                        |
   |                        immutable borrow occurs here
   |                        immutable borrow later used here
52 |                 self.send_prepare(*pid);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here

```

Reply:
Ok since this is in the prepare phase i.e., it is performed rarely it should be fine.